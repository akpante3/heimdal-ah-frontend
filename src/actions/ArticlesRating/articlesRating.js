import instance from '../../config/http';
import {
  ACTIONS,
} from '../actionTypes';
import ActionResponse from '../actionResponse';

const articleRating = (stars, articleId) => dispatch => instance.post(`/ratings/articles/${articleId}`, {
  stars,
}).then((response) => {
  if (response.status === 201) {
    dispatch(ActionResponse(ACTIONS.ARTICLES_RATING_SUCCESS, response.data.ratings.stars));
    dispatch(ActionResponse(ACTIONS.RESET_ARTICLES_RATING_STATE));
  }
}).catch((error) => {
  dispatch(ActionResponse(ACTIONS.ARTICLES_RATING_ERROR, 'Server error', error));
  dispatch(ActionResponse(ACTIONS.RESET_ARTICLES_RATING_STATE));
});

export default articleRating;
