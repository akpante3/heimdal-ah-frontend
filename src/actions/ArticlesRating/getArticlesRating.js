import instance from '../../config/http';
import {
  ACTIONS,
} from '../actionTypes';
import ActionResponse from '../actionResponse';

const getAticleRating = articleId => dispatch => instance.get(`/articles/${articleId}`).then((response) => {
  if (response.status === 200) {
    dispatch(ActionResponse(ACTIONS.GET_ARTICLES_RATING_SUCCESS, response.data.article.ratings));
  }
}).catch((error) => {
  dispatch(ActionResponse(ACTIONS.GET_ARTICLES_RATING_ERROR, 'Server error', error));
});

export default getAticleRating;
