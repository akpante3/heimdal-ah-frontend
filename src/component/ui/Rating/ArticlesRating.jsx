import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { range } from 'lodash';
import './Rating.scss';
import articleRating from '../../../actions/ArticlesRating/articlesRating';
import getAverageRatings from '../../../utils/getAverageRatings';
import getArticleRating from '../../../actions/ArticlesRating/getArticlesRating';
import Alert from '../Alert/Alert';

/**
 * @description - Helps a user rate an article
 * @param {props} status - the status returned from dispatching articleRating action
 * @param {props} star - the star rating given by a user to the article
 * @param {props} getratingstatus - the status returned from dispatching getArticleRating action
 * @param {props} ratings - average rating for the article
 * @returns {component} Component
 */
export class ArticlesRating extends Component {
  constructor(props) {
    super(props);
    this.onHandleClick = this.onHandleSubmit.bind(this);
  }

  /**
 * @description - Handles the initial component mounting and calls getArticleRating action
 */
  componentDidMount() {
    const articleId = setTimeout((localStorage.getItem('articleId')), 1);
    this.props.actions(getArticleRating(articleId));
  }

  /**
 * @description - Handles the submission of the rating by calling articleRating action
 * @param {number} stars - the star rating of the article, supplied by user
 */
  onHandleSubmit(stars) {
    // Might change based on where the article id is  gotten from
    const articleId = localStorage.getItem('articleId');
    // eslint-disable-next-line react/destructuring-assignment
    this.props.actions(articleRating(stars, articleId));
  }

  /**
 * @description - Handles the rendering of the component
 * @returns {component} the rendered component
 */
  render() {
    return (
      <Fragment>
        {this.props.status === 'SUCCESS' && <Alert type='success' title='Article Rated' message={`Hello Buddy, you gave this article a star rating of ${this.props.star}`} />}
        {this.props.getratingstatus === 'SUCCESS' && (
        <span className='rate'>
          {
            range(1, 6).map(value => (
              <button
                key={value}
                type='button'
                onClick={() => {
                  this.onHandleSubmit(value);
                }}
                className='rate __button'
              >
                <i
                  key={value}
                  className={`${ArticlesRating.getClassNames(value, getAverageRatings(Array.from(this.props.ratings)))}  `}
                />
              </button>
            ))
          }
        </span>
        )
        }
      </Fragment>
    );
  }
}

ArticlesRating.getClassNames = (value, ratings) => (
  (value <= Math.trunc(ratings)) ? 'fa fa-star star' : 'fa fa-star-o star'
);

ArticlesRating.defaultProps = {
  ratings: 0,
};

const mapStateToProps = state => ({
  status: state.articlerating.status,
  star: state.articlerating.payload,
  ratings: state.getarticlerating.payload,
  getratingstatus: state.getarticlerating.status,
});

const mapDispatchToProps = dispatch => ({
  actions: action => dispatch(action),
});

export default connect(mapStateToProps, mapDispatchToProps)(ArticlesRating);
