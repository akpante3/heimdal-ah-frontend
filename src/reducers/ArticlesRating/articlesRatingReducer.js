import {
  ACTIONS,
} from '../../actions/actionTypes';

const initialState = {
  payload: '',
  status: '',
};

export default (state = initialState, action) => {
  switch (action.type) {
    case ACTIONS.RESET_ARTICLES_RATING_STATE:
      return ({
        ...state,
        payload: '',
        status: '',
      });

    case ACTIONS.ARTICLES_RATING_SUCCESS:
      return ({
        ...state,
        payload: action.payload,
        status: 'SUCCESS',
      });

    case ACTIONS.ARTICLES_RATING_ERROR:
      return ({
        ...state,
        payload: action.payload,
        status: 'ERROR',
      });

    default:
      return (state);
  }
};
