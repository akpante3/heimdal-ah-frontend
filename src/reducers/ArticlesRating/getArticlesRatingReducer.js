import {
  ACTIONS,
} from '../../actions/actionTypes';

const initialState = {
  payload: '',
  status: '',
};

export default (state = initialState, action) => {
  switch (action.type) {
    case ACTIONS.GET_ARTICLES_RATING_SUCCESS:
      return ({
        ...state,
        payload: action.payload,
        status: 'SUCCESS',
      });

    case ACTIONS.GET_ARTICLES_RATING_ERROR:
      return ({
        ...state,
        payload: action.payload,
        status: 'ERROR',
      });

    default:
      return (state);
  }
};
