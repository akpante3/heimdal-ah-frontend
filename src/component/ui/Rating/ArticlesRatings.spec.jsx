/* eslint-disable no-undef */
import React from 'react';
import { ArticlesRating } from './ArticlesRating';

describe('Render <ArticleRating />', () => {
  const wrapper = shallow(<ArticlesRating actions={() => { }} getratingstatus='SUCCESS' status='SUCCESS' />);
  it('it should render  ArticleRating component', () => {
    expect(wrapper.length).toBe(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('it should simulate click event', () => {
    const spy = jest.spyOn(ArticlesRating.prototype, 'onHandleSubmit');
    wrapper.find('button').at(0).simulate('click');
    expect(spy).toHaveBeenCalled();
  });

  it('should have font awesome class', () => {
    expect(wrapper.find('i').at(0).hasClass('fa fa-star star'));
  });
});
