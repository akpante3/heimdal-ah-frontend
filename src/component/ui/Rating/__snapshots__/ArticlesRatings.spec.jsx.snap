// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render <ArticleRating /> it should render  ArticleRating component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ArticlesRating
    actions={[Function]}
    getratingstatus="SUCCESS"
    ratings={0}
    status="SUCCESS"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Alert
          message="Hello Buddy, you gave this article a star rating of undefined"
          title="Article Rated"
          type="success"
        />,
        <span
          className="rate"
        >
          <button
            className="rate __button"
            onClick={[Function]}
            type="button"
          >
            <i
              className="fa fa-star-o star  "
            />
          </button>
          <button
            className="rate __button"
            onClick={[Function]}
            type="button"
          >
            <i
              className="fa fa-star-o star  "
            />
          </button>
          <button
            className="rate __button"
            onClick={[Function]}
            type="button"
          >
            <i
              className="fa fa-star-o star  "
            />
          </button>
          <button
            className="rate __button"
            onClick={[Function]}
            type="button"
          >
            <i
              className="fa fa-star-o star  "
            />
          </button>
          <button
            className="rate __button"
            onClick={[Function]}
            type="button"
          >
            <i
              className="fa fa-star-o star  "
            />
          </button>
        </span>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "message": "Hello Buddy, you gave this article a star rating of undefined",
          "title": "Article Rated",
          "type": "success",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              className="rate __button"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-star-o star  "
              />
            </button>,
            <button
              className="rate __button"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-star-o star  "
              />
            </button>,
            <button
              className="rate __button"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-star-o star  "
              />
            </button>,
            <button
              className="rate __button"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-star-o star  "
              />
            </button>,
            <button
              className="rate __button"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-star-o star  "
              />
            </button>,
          ],
          "className": "rate",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "host",
            "props": Object {
              "children": <i
                className="fa fa-star-o star  "
              />,
              "className": "rate __button",
              "onClick": [Function],
              "type": "button",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": "1",
              "nodeType": "host",
              "props": Object {
                "className": "fa fa-star-o star  ",
              },
              "ref": null,
              "rendered": null,
              "type": "i",
            },
            "type": "button",
          },
          Object {
            "instance": null,
            "key": "2",
            "nodeType": "host",
            "props": Object {
              "children": <i
                className="fa fa-star-o star  "
              />,
              "className": "rate __button",
              "onClick": [Function],
              "type": "button",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": "2",
              "nodeType": "host",
              "props": Object {
                "className": "fa fa-star-o star  ",
              },
              "ref": null,
              "rendered": null,
              "type": "i",
            },
            "type": "button",
          },
          Object {
            "instance": null,
            "key": "3",
            "nodeType": "host",
            "props": Object {
              "children": <i
                className="fa fa-star-o star  "
              />,
              "className": "rate __button",
              "onClick": [Function],
              "type": "button",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": "3",
              "nodeType": "host",
              "props": Object {
                "className": "fa fa-star-o star  ",
              },
              "ref": null,
              "rendered": null,
              "type": "i",
            },
            "type": "button",
          },
          Object {
            "instance": null,
            "key": "4",
            "nodeType": "host",
            "props": Object {
              "children": <i
                className="fa fa-star-o star  "
              />,
              "className": "rate __button",
              "onClick": [Function],
              "type": "button",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": "4",
              "nodeType": "host",
              "props": Object {
                "className": "fa fa-star-o star  ",
              },
              "ref": null,
              "rendered": null,
              "type": "i",
            },
            "type": "button",
          },
          Object {
            "instance": null,
            "key": "5",
            "nodeType": "host",
            "props": Object {
              "children": <i
                className="fa fa-star-o star  "
              />,
              "className": "rate __button",
              "onClick": [Function],
              "type": "button",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": "5",
              "nodeType": "host",
              "props": Object {
                "className": "fa fa-star-o star  ",
              },
              "ref": null,
              "rendered": null,
              "type": "i",
            },
            "type": "button",
          },
        ],
        "type": "span",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Alert
            message="Hello Buddy, you gave this article a star rating of undefined"
            title="Article Rated"
            type="success"
          />,
          <span
            className="rate"
          >
            <button
              className="rate __button"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-star-o star  "
              />
            </button>
            <button
              className="rate __button"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-star-o star  "
              />
            </button>
            <button
              className="rate __button"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-star-o star  "
              />
            </button>
            <button
              className="rate __button"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-star-o star  "
              />
            </button>
            <button
              className="rate __button"
              onClick={[Function]}
              type="button"
            >
              <i
                className="fa fa-star-o star  "
              />
            </button>
          </span>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "message": "Hello Buddy, you gave this article a star rating of undefined",
            "title": "Article Rated",
            "type": "success",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                className="rate __button"
                onClick={[Function]}
                type="button"
              >
                <i
                  className="fa fa-star-o star  "
                />
              </button>,
              <button
                className="rate __button"
                onClick={[Function]}
                type="button"
              >
                <i
                  className="fa fa-star-o star  "
                />
              </button>,
              <button
                className="rate __button"
                onClick={[Function]}
                type="button"
              >
                <i
                  className="fa fa-star-o star  "
                />
              </button>,
              <button
                className="rate __button"
                onClick={[Function]}
                type="button"
              >
                <i
                  className="fa fa-star-o star  "
                />
              </button>,
              <button
                className="rate __button"
                onClick={[Function]}
                type="button"
              >
                <i
                  className="fa fa-star-o star  "
                />
              </button>,
            ],
            "className": "rate",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "host",
              "props": Object {
                "children": <i
                  className="fa fa-star-o star  "
                />,
                "className": "rate __button",
                "onClick": [Function],
                "type": "button",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": "1",
                "nodeType": "host",
                "props": Object {
                  "className": "fa fa-star-o star  ",
                },
                "ref": null,
                "rendered": null,
                "type": "i",
              },
              "type": "button",
            },
            Object {
              "instance": null,
              "key": "2",
              "nodeType": "host",
              "props": Object {
                "children": <i
                  className="fa fa-star-o star  "
                />,
                "className": "rate __button",
                "onClick": [Function],
                "type": "button",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": "2",
                "nodeType": "host",
                "props": Object {
                  "className": "fa fa-star-o star  ",
                },
                "ref": null,
                "rendered": null,
                "type": "i",
              },
              "type": "button",
            },
            Object {
              "instance": null,
              "key": "3",
              "nodeType": "host",
              "props": Object {
                "children": <i
                  className="fa fa-star-o star  "
                />,
                "className": "rate __button",
                "onClick": [Function],
                "type": "button",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": "3",
                "nodeType": "host",
                "props": Object {
                  "className": "fa fa-star-o star  ",
                },
                "ref": null,
                "rendered": null,
                "type": "i",
              },
              "type": "button",
            },
            Object {
              "instance": null,
              "key": "4",
              "nodeType": "host",
              "props": Object {
                "children": <i
                  className="fa fa-star-o star  "
                />,
                "className": "rate __button",
                "onClick": [Function],
                "type": "button",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": "4",
                "nodeType": "host",
                "props": Object {
                  "className": "fa fa-star-o star  ",
                },
                "ref": null,
                "rendered": null,
                "type": "i",
              },
              "type": "button",
            },
            Object {
              "instance": null,
              "key": "5",
              "nodeType": "host",
              "props": Object {
                "children": <i
                  className="fa fa-star-o star  "
                />,
                "className": "rate __button",
                "onClick": [Function],
                "type": "button",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": "5",
                "nodeType": "host",
                "props": Object {
                  "className": "fa fa-star-o star  ",
                },
                "ref": null,
                "rendered": null,
                "type": "i",
              },
              "type": "button",
            },
          ],
          "type": "span",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
